File: frontend/src/features/racing/lanes.ts
/**
 * Predicates over a heat's lanes.
 *
 * These questions — has this heat run? was it skipped? — were asked in eight
 * places across the race-control screens, each with its own inline copy of the
 * test, and they had drifted: some counted a skipped heat as run and some did
 * not. Naming them makes the difference deliberate rather than accidental.
 *
 * Also the conversion to `HeatLaneInput` for the write path, which is a
 * near-identity — the read and write shapes match on purpose.
 */
import type { Heat, Lane, LaneInput } from './types';

/**
 * A lane with a recorded result — a time, or a hand-entered place.
 *
 * "Time" is the common case and the name predates the other one. A `POINTS`
 * race entered by hand through the Override/Edit modal — no timer, or a
 * timer that only reports finishing order (#490) — writes a place with no
 * time at all, and that is a result too. Mirrors `Lane.has_result` in
 * `backend/domain/lanes.py`: before the modal could enter a place on its
 * own, a lane never held one without a time, so broadening this changes
 * nothing for data recorded before #490.
 */
export const hasTime = (lane: Lane): boolean => lane.time !== null || lane.place !== null;

/** Any result recorded in this heat. */
export const hasTimes = (lanes: readonly Lane[]): boolean => lanes.some(hasTime);

/**
 * The heat is done with — raced, or passed over.
 *
 * The operator can skip a heat (everyone in it scratched, say), and for
 * "what's next" purposes that is as finished as one that ran. Note the backend
 * disagrees: `lanes.has_results` ignores `skipped`, so a skipped round can
 * still be regenerated.
 */
export const hasRun = (lanes: readonly Lane[]): boolean =>
  lanes.some((lane) => hasTime(lane) || lane.skipped);

/** Passed over rather than raced — skipped, and nothing was timed. */
export const wasSkipped = (lanes: readonly Lane[]): boolean =>
  lanes.some((lane) => lane.skipped) && !hasTimes(lanes);

/**
 * Heats that were skipped and never re-run, in heat-number order (issue
 * #1001) — the earliest is the one an operator would want to run first.
 * `wasSkipped` already asks the per-heat question ("skipped, and nothing has
 * been timed since"); this lifts it to a list of heats for the Round
 * Complete!/Race Complete! summaries, both of which need to name *which*
 * heat rather than just whether one exists.
 */
export const skippedHeats = (heats: readonly Heat[]): Heat[] =>
  heats.filter((h) => wasSkipped(h.lanes)).sort((a, b) => a.heatNumber - b.heatNumber);

/** Lanes in finishing order, unplaced last. */
export const byPlace = (lanes: readonly Lane[]): Lane[] =>
  [...lanes].sort((a, b) => (a.place ?? 99) - (b.place ?? 99));

/**
 * The racer in this lane, if it holds one.
 *
 * `null` covers both an empty lane and a championship slot whose racer has not
 * been decided yet — `placeholderSlot` tells those apart when it matters.
 */
export const racerIdIn = (lane: Lane): number | null => lane.racerId;

/**
 * A lane as the mutation takes it.
 *
 * Field-for-field the same as {@link Lane}, but spelt out rather than spread:
 * the cache attaches `__typename` to what it hands back, and GraphQL rejects an
 * input object carrying a field the type does not declare.
 */
export const toInput = (lane: Lane): LaneInput => ({
  lane: lane.lane,
  racerId: lane.racerId,
  placeholderSlot: lane.placeholderSlot,
  time: lane.time,
  place: lane.place,
  skipped: lane.skipped,
});

/** The same lanes with any result removed — what re-running a heat sends. */
export const cleared = (lanes: readonly Lane[]): LaneInput[] =>
  lanes.map((lane) => ({ ...toInput(lane), time: null, place: null, skipped: false }));

/**
 * Stamp finishing places over a heat's edited results.
 *
 * Mirrors `backend/domain/scoring.py`'s DNF rule: a recorded time of zero or
 * less is not a finish — the timer assigns it no place, and `POINTS` scores
 * it as a last-place penalty rather than as a placement (issue #308). Ranking
 * every recorded time, including a DNF's `0.0`, undid that: the DNF lane
 * sorted first and was stamped `place = 1`.
 *
 * Only lanes with a real time (`time > 0`) are ranked; a null, zero or
 * negative time gets `place: null`, same as an unrun lane.
 *
 * A heat with no recorded time at all (the operator hit Skip) clears every
 * place without touching `skipped` — this is not that heat's concern, since
 * the caller has already set `skipped` on the lanes it wants marked.
 */
export const assignPlaces = (results: readonly LaneInput[]): LaneInput[] => {
  const hasAnyTime = results.some((r) => r.time !== null);
  if (!hasAnyTime) {
    return results.map((r) => ({ ...r, place: null }));
  }

  const finishers = results
    .filter((r): r is LaneInput & { time: number } => typeof r.time === 'number' && r.time > 0)
    .sort((a, b) => a.time - b.time);
  const placeByLane = new Map<number, number>();
  let currentPlace = 1;
  for (let i = 0; i < finishers.length; i++) {
    if (i > 0 && finishers[i].time > finishers[i - 1].time) {
      currentPlace = i + 1;
    }
    placeByLane.set(finishers[i].lane, currentPlace);
  }

  return results.map((r) => ({
    ...r,
    skipped: false, // Always clear skipped flag if we have any time.
    place: placeByLane.get(r.lane) ?? null,
  }));
};

/**
 * A per-lane heat time, formatted for display (#763). Every screen that
 * shows a raw recorded time — RaceExecution's live lane cards,
 * ScheduleManagement's and RaceControl's schedule tables, FreeRaceExecution,
 * the projector and broadcast overlay, and the timer diagnostics page — had
 * its own `.toFixed()` call, three decimals on most of them and four on the
 * live-heat screens, with nothing saying why the fourth digit was there. It
 * was not a deliberate finer reading: it is one rule now, three decimals
 * everywhere, matching what every real timer profile's own precision
 * supports and what the standings score formatter
 * (`scoringStrategyText.formatScore`) already uses.
 *
 * `null`/`undefined` (nothing recorded yet) returns `null` rather than a
 * fallback string, since that varies by screen ("--", "—", blank); the
 * caller supplies its own with `formatLaneTime(time) ?? '—'`.
 *
 * A time at or below zero is the DNF marker this file's own `assignPlaces`
 * already reads, mirroring `backend/domain/scoring.py`: the timer assigns it
 * no place, and it scores as a penalty rather than a real result. Printing
 * "0.000s" — or, worse, a bare `9.999s` penalty value with nothing marking
 * it as one — is how an operator mistakes a scratch for a result; this
 * always prints "DNF" for a non-positive time instead of the number.
 */
export const formatLaneTime = (time: number | null | undefined): string | null => {
  if (time == null) {
    return null;
  }
  if (time <= 0) {
    return 'DNF';
  }
  return `${time.toFixed(3)}s`;
};

/**
 * Whether `scoringStrategy` scores from a recorded *time* rather than a
 * hand-entered *place* — the distinction the Override/Edit modal's column
 * choice turns on (#490), and the one thing #547's two new strategies have
 * to get right on the frontend.
 *
 * `POINTS` is the only place-based strategy today, so this is `!== 'POINTS'`
 * — but it is stated as its own predicate, not spelled out at each call
 * site, because every strategy #547 adds (`CUMULATIVE_TIME`,
 * `FASTEST_TIME`) is time-based and a future member that sums *placements*
 * the way `POINTS` does would need exactly one place updated rather than
 * every site that copied the "not `POINTS`" test agreeing with it by luck.
 * Mirrors `backend.domain.scoring`'s module docstring, which states the same
 * four strategies.
 */
export const isTimeBasedStrategy = (scoringStrategy: string | null | undefined): boolean =>
  scoringStrategy !== 'POINTS';

/**
 * Whether saving an official heat's edited results should run them through
 * {@link assignPlaces} (issue #490).
 *
 * Mirrors which column the Override/Edit modal requires: a `TIMED` race
 * enters times and always wants them turned into places — the rule
 * `FreeRaceExecution` already follows unconditionally, since free racing has
 * no place column to enter by hand. A `POINTS` race enters places directly,
 * with no time to derive them from — `assignPlaces` reads "no time anywhere"
 * as "clear every place", which is exactly backwards for a lane the operator
 * just placed by hand, so it must not run at all. The modal shows a Time
 * column for a `POINTS` race too (#525), so a stored or spurious time stays
 * correctable, but that column is optional there and must never drive
 * derivation the way it does for `TIMED`.
 *
 * Deciding this from the strategy rather than from the edited lanes' own
 * shape (e.g. "times present and places absent") matters for a *correction*:
 * re-editing an already-placed `TIMED` heat to fix one mistyped time must
 * still recompute every place from the new times, even though the old places
 * are sitting right there in the payload.
 */
export const shouldDerivePlaces = (scoringStrategy: string | null | undefined): boolean =>
  isTimeBasedStrategy(scoringStrategy);

/**
 * Whether saving a free-race heat's edited results should run them through
 * {@link assignPlaces} (issue #526).
 *
 * `shouldDerivePlaces` keys off the race's scoring strategy, which is the
 * wrong question here: a free heat is exhibition and is excluded from
 * scoring under either strategy (#6). What decides it is whether the
 * *track* has a timer (#490's `hasTimer`) — with one, the modal still only
 * takes times and always wants them turned into places; with none, the
 * modal takes a hand-typed finishing order directly, and `assignPlaces`
 * reading "no time anywhere" as "clear every place" would erase exactly
 * what the operator just typed.
 */
export const shouldDerivePlacesForFreeRace = (hasTimer: boolean): boolean => hasTimer;

/**
 * Real racers assigned a lane — an empty lane or an undecided championship
 * slot is not a competitor to place behind. Mirrors
 * `backend/domain/lanes.py`'s `real_racer_ids`.
 */
const realRacerCount = (results: readonly LaneInput[]): number =>
  results.filter((r) => r.racerId !== null).length;

/**
 * Lane numbers whose hand-entered place is not a positive number (issue
 * #766, extending #524's server-side rule to the client). `handleResultChange`
 * in `RaceExecution.tsx` already refuses to store one of these as the
 * operator types, so this mostly documents that the invariant holds rather
 * than catching anything live — but it is what makes {@link placeIssue} a
 * complete mirror of `crud.validate_lane_replacement` rather than an
 * incomplete one somebody has to remember not to trust. Mirrors
 * `backend/domain/lanes.py`'s `places_below_one`.
 */
export const placesBelowOne = (results: readonly LaneInput[]): number[] =>
  results.filter((r) => r.place != null && r.place < 1).map((r) => r.lane);

/**
 * Lane numbers whose place exceeds the number of real racers in the heat
 * (issue #766). Mirrors `backend/domain/lanes.py`'s `places_above_field` —
 * see there for why an empty field checks nothing.
 */
export const placesAboveField = (results: readonly LaneInput[]): number[] => {
  const field = realRacerCount(results);
  if (!field) return [];
  return results.filter((r) => r.place != null && r.place > field).map((r) => r.lane);
};

/**
 * Place values claimed by more than one lane, each named once (issue #766).
 * Mirrors `backend/domain/lanes.py`'s `duplicate_places`.
 *
 * A genuine tie in recorded times — multiple lanes sharing the identical
 * positive finish time — genuinely shares a place and is not flagged (#816).
 * Duplicate places without matching times are flagged.
 */
export const duplicatePlaces = (results: readonly LaneInput[]): number[] => {
  const byPlace = new Map<number, LaneInput[]>();
  for (const r of results) {
    if (r.place == null) continue;
    const group = byPlace.get(r.place) ?? [];
    group.push(r);
    byPlace.set(r.place, group);
  }

  const dupes: number[] = [];
  for (const [place, group] of byPlace.entries()) {
    if (group.length > 1) {
      const firstTime = group[0].time;
      const isGenuineTie =
        firstTime !== null &&
        firstTime !== undefined &&
        typeof firstTime === 'number' &&
        firstTime > 0 &&
        group.every((r) => r.time === firstTime);

      if (!isGenuineTie) {
        dupes.push(place);
      }
    }
  }
  return dupes;
};

/**
 * The first problem with a hand-entered set of places, or `null` if there
 * isn't one (issue #766).
 *
 * Mirrors `crud.validate_lane_replacement`'s own place checks — same order,
 * same wording — so the client catches what the server would refuse before
 * the operator ever clicks Save, and reads the identical sentence in the
 * rare case something still reaches the server unchecked (a stale prop, a
 * second tab). It is a *first* check, not a replacement one: the server
 * remains the backstop for anything this can't see, e.g. a lane set that
 * does not match the heat's own schedule.
 *
 * Deliberately does not check lane numbers themselves (duplicate or unknown
 * lanes) — those come from the heat's own stored schedule, never from
 * anything the operator types, so there is nothing here to mistype.
 */
export const placeIssue = (results: readonly LaneInput[]): string | null => {
  const belowOne = placesBelowOne(results);
  if (belowOne.length) return `Lane ${belowOne[0]}'s place must be 1 or higher.`;

  const aboveField = placesAboveField(results);
  if (aboveField.length) {
    const field = realRacerCount(results);
    return `Lane ${aboveField[0]}'s place is higher than the ${field} racer(s) in this heat.`;
  }

  const dupes = duplicatePlaces(results);
  if (dupes.length) return `Place ${dupes[0]} is assigned to more than one lane.`;

  return null;
};

/**
 * A typed time field as the number it will be saved as, or `null` for
 * blank or unparsable text — factored out of `RaceExecution.tsx`'s
 * `handleSaveResults` (issue #766) so the equal-time tie detector below
 * reads the exact value that will be saved, rather than a second copy of
 * the same parsing rule free to drift from it.
 */
export const parseTimeText = (timeText: string): number | null => {
  const time = Number(timeText);
  return timeText.trim() === '' || isNaN(time) ? null : time;
};

/**
 * How many lane columns a schedule table or heat sheet needs (issue #994).
 *
 * `track.laneCount` is usually enough, but a track is shared across seasons
 * (`.claude/rules/scheduling.md`): shrinking it after a race leaves that
 * race's finished heats naming a lane the track no longer has. #325 rewrites
 * *pending* heats to match a new lane count, deliberately not recorded
 * ones — the data is real and must not be discarded — so a screen that
 * renders exactly `laneCount` columns silently clips that lane's results off
 * the page, even though `heats { lanes { lane time } }` still holds them.
 *
 * The fix is not "read the heats instead of the track" — that would break
 * the rule this one extends: a heat short a lane (an outage, say) must still
 * line up with its neighbours, showing an empty column rather than shifting
 * left, which only works if every row has at least `laneCount` columns. So
 * this is a `max`: at least the track's own lanes, and at least whatever
 * lane number the given heats actually hold.
 *
 * One function for both callers (`ScheduleManagement.tsx`,
 * `HeatSheet.tsx`) so they cannot independently derive the column count and
 * disagree about it the way they did before this existed.
 */
export const laneColumnCount = (
  laneCount: number,
  heats: readonly { lanes: readonly { lane: number }[] }[],
): number =>
  heats.reduce(
    (max, heat) => heat.lanes.reduce((laneMax, l) => Math.max(laneMax, l.lane), max),
    laneCount,
  );

/**
 * Lanes that recorded the identical time, grouped by that time (issue
 * #766). `assignPlaces` breaks a genuine tie by array order with nothing on
 * screen to say a tie happened — two identical hand-typed times silently
 * become 2nd and 3rd. This does not change what gets saved: for a
 * `TIMED` race there is no Place column to correct it from, and rewriting
 * the times themselves is not this function's business. It exists so
 * `RaceExecution` can show the operator the tie exists, the same "computed,
 * never rewritten" shape {@link placeIssue} follows for the Place column.
 *
 * Mirrors `assignPlaces`'s own finisher filter — only a real time (`> 0`)
 * counts, so a DNF's `0` or a negative marker never reads as a three-way
 * tie with every other DNF in the heat.
 */
export const tiedTimeGroups = (
  results: readonly { lane: number; time: number | null }[],
): { time: number; lanes: number[] }[] => {
  const byTime = new Map<number, number[]>();
  for (const r of results) {
    if (r.time === null || r.time <= 0) continue;
    const group = byTime.get(r.time) ?? [];
    group.push(r.lane);
    byTime.set(r.time, group);
  }
  return [...byTime.entries()]
    .filter(([, lanes]) => lanes.length > 1)
    .map(([time, lanes]) => ({ time, lanes: [...lanes].sort((a, b) => a - b) }))
    .sort((a, b) => a.time - b.time);
};

File: backend/domain/lanes.py
"""The lane assignment/result value object.

:class:`Lane` is what the domain layer passes around — scheduling, scoring,
advancement and ``heat_session`` all take parsed lanes — and what
``crud.set_heat_lanes`` stores. It carries who is in a lane, their time and
place, and whether the lane is an unadvanced championship slot.

This module used to be the codec for ``Heat.lane_results`` as well: a JSON
string doing four jobs at once, and the only place that knew it. #72 finished
replacing that with the ``heat_lanes`` table, so ``parse``, ``serialize``,
``from_dict``, ``to_dict`` and ``carry_extras`` are gone. What is left is the
value and the predicates over it.

Two of the blob's conventions outlived it and are now gone too (#164). An
unadvanced championship slot was a *negative* ``racer_id`` — slot 1 was ``-1`` —
and ``skipped`` lived in an ``extra`` dict, because a blob could carry keys
nothing modelled. Both are fields now, and ``extra`` is gone: ``heat_lanes`` has
a column for everything, and ``HeatLaneInput`` is typed, so no unknown key can
arrive.

One consequence worth stating, because it is the trap in that change:
:attr:`Lane.is_empty` asks about *both* fields. A placeholder used to hold a
negative id, so it was never "empty"; with the id now ``None`` it would be,
and :func:`is_complete` would skip it and call a round of undecided slots
finished.

``time`` is stored as it was found. ``heat_lanes.time_seconds`` is a float, so
a string can no longer be *persisted*, but the field is still :class:`Any` and
:attr:`Lane.seconds` is the coercion every reader goes through.
"""

from __future__ import annotations

from collections.abc import Iterable, Sequence
from dataclasses import dataclass
from typing import Any


@dataclass
class Lane:
    """One lane of one heat: an assignment, and possibly a result."""

    lane: int
    #: The racer in this lane. ``None`` for an unused lane *and* for a
    #: championship slot nobody has advanced into yet — see
    #: :attr:`placeholder_slot`, which is what tells those two apart.
    racer_id: int | None = None
    #: Which championship slot this lane holds, 1-based, or ``None``.
    placeholder_slot: int | None = None
    time: Any = None
    place: int | None = None
    #: Set by the operator UI when a heat is passed over rather than run.
    skipped: bool = False

    @property
    def is_empty(self) -> bool:
        """No racer and no slot — a short heat's unused lane.

        Both fields, deliberately. An unadvanced slot has no ``racer_id`` and is
        emphatically not an unused lane: :func:`is_complete` skips empty lanes,
        so calling one empty would let a round of undecided slots read as
        finished.
        """
        return self.racer_id is None and self.placeholder_slot is None

    @property
    def is_placeholder(self) -> bool:
        """An unadvanced championship slot."""
        return self.placeholder_slot is not None

    @property
    def has_result(self) -> bool:
        """This lane holds a result — a time, or a hand-entered place.

        Historically a lane never held a ``place`` without a ``time``: every
        writer that set one set both, because the only way a place ever got
        there was :func:`backend.domain.scoring` deriving it from a time.
        Issue #490 adds the first exception — a ``POINTS`` race entered by
        hand through the Override/Edit modal writes a place with no time at
        all, because there is nothing to time. That is a result too, so this
        broadened to match rather than leaving every caller that asks "has
        this lane been decided" — :func:`has_results`, :func:`is_finished`,
        ``heat_session.is_recorded`` — blind to it. Data recorded before
        #490 never had a place without a time, so this changes nothing for
        it.
        """
        return self.time is not None or self.place is not None

    @property
    def seconds(self) -> float | None:
        """``time`` as a number, or ``None`` if it is absent or unparseable.

        Does not apply the DNF penalty — that is a scoring decision and lives in
        :mod:`backend.domain.scoring`.
        """
        if self.time is None:
            return None
        try:
            return float(self.time)
        except (TypeError, ValueError):
            return None


def from_parts(
    *,
    lane: int,
    racer_id: int | None,
    placeholder_slot: int | None,
    time_seconds: float | None,
    place: int | None,
    skipped: bool,
) -> Lane:
    """A lane from the columns ``heat_lanes`` stores it in (#72).

    The inverse of the projection in ``db/lane_sync.py``, and the read path's
    entry point. Scalars rather than a row object, so this module still imports
    no SQLAlchemy — that is the whole reason it exists now that the fields line
    up one for one.
    """
    return Lane(
        lane=lane,
        racer_id=racer_id,
        placeholder_slot=placeholder_slot,
        time=time_seconds,
        place=place,
        skipped=skipped,
    )


def from_participant(lane: int, participant_id: int | None) -> Lane:
    """A scheduled lane, from the id the scheduler deals in.

    :mod:`backend.domain.scheduling` matches *opaque* ids — it neither knows
    nor cares whether one is a racer — and its ``placeholder_ids`` hands it
    negative ones for championship slots nobody has advanced into yet.
    That is the scheduler's vocabulary, and it is a good one: teaching a matching
    algorithm about advancement would be worse than translating at its edge.

    This is that edge, and the only place on the write path that knows a
    negative participant id means a slot (#164).
    """
    if participant_id is not None and participant_id < 0:
        return Lane(lane=lane, placeholder_slot=-participant_id)
    return Lane(lane=lane, racer_id=participant_id)


def has_results(lanes: Sequence[Lane]) -> bool:
    """True if any lane has a recorded result — a time, or a hand-entered place.

    This is the check that guards regeneration: a round that has been raced must
    not be silently rebuilt underneath the operator. A ``POINTS`` round entered
    by hand (#490) has no times at all, and still must not be rebuildable.

    Note it does **not** consider :attr:`Lane.skipped` — see :func:`is_finished`
    for the predicate that does. Making this one stricter would change when
    regeneration is refused: a skipped round holds no times, so there is nothing
    to lose by rebuilding it.
    """
    return any(lane.has_result for lane in lanes)


def is_finished(lanes: Sequence[Lane]) -> bool:
    """True if the race is done with this heat — raced, or passed over.

    The counterpart to :func:`has_results`, and the difference between them is
    the point. ``has_results`` asks "would rebuilding this lose a result", which
    is a question about the stored record. This asks "should the displays move
    on", which is a question about the running order: an operator who skipped a
    heat is not coming back to it.

    Matches ``hasRun`` in ``features/racing/lanes.ts``. Anything deciding what
    is on the track or what is next wants this one; #55 is what happens when it
    settles for the other.
    """
    return any(lane.has_result or lane.skipped for lane in lanes)


def is_complete(lanes: Sequence[Lane]) -> bool:
    """True if every assigned lane is *settled* — a time, a place, or skipped.

    This is the predicate behind ``is_round_complete``, which is what decides
    when championship advancement fires, so what counts as settled matters
    (#224). It used to demand a time on every lane, which was wrong twice over:

    * A **skipped** heat has no times, so one skip left its round incomplete
      forever — the automatic advancement never fired and ``advancementStatus``
      reported not-ready with no way to learn why. An operator who skipped a
      heat is not coming back to it; the round is as decided as it will ever be.
    * A **place without a time** is how a ``POINTS`` race is entered by hand.
      ``advancementStatus`` grew a private copy of this rule to accept that,
      which meant the operator screen said ready while the trigger reading
      *this* function never fired. One rule, two copies, two answers — the #48
      shape. The copy is gone; this is the rule.

    A placeholder counts as incomplete even if it somehow has a time: the heat
    cannot be finished by a racer who has not been decided yet.
    """
    if not lanes:
        return False
    for lane in lanes:
        if lane.is_empty:
            continue
        if lane.is_placeholder:
            return False
        if lane.time is None and lane.place is None and not lane.skipped:
            return False
    return True


def real_racer_ids(lanes: Iterable[Lane]) -> list[int]:
    """Assigned racer ids, in lane order.

    Still worth a name: it is dense, so it drops unused lanes and undecided
    slots rather than yielding ``None`` for them.
    """
    return [lane.racer_id for lane in lanes if lane.racer_id is not None]


def duplicate_lane_numbers(lanes: Sequence[Lane]) -> list[int]:
    """Lane numbers claimed by more than one row, each named once.

    A replacement lane set is a client's whole description of a heat
    (:func:`backend.db.crud.validate_lane_replacement`); two rows both
    claiming lane 1 is not a heat, it is two heats squashed into one payload,
    and storing it silently drops one of them. Named in the order they first
    repeat, so the error can point at the first offender rather than an
    unordered set.
    """
    seen: set[int] = set()
    dupes: list[int] = []
    for lane in lanes:
        if lane.lane in seen:
            if lane.lane not in dupes:
                dupes.append(lane.lane)
        else:
            seen.add(lane.lane)
    return dupes


def resolve_placeholders(lanes: Sequence[Lane], racer_ids: Sequence[int]) -> bool:
    """Replace placeholder slots with real racers, in place.

    Placeholder ``-1`` takes ``racer_ids[0]``, ``-2`` takes ``racer_ids[1]``,
    and so on. A placeholder with no corresponding entry is left alone, which is
    what happens when fewer racers advanced than the round has slots.

    Returns whether anything changed, so callers can skip a needless write.
    """
    modified = False
    for lane in lanes:
        slot = lane.placeholder_slot
        if slot is None:
            continue
        index = slot - 1
        if index < len(racer_ids):
            lane.racer_id = racer_ids[index]
            # Clearing this is not tidying. While a slot was a negative id,
            # writing the racer over it *was* the clear; with two fields, a lane
            # that keeps its slot stays a placeholder however real its racer is,
            # and `phase` reports NOT_READY for a round that is ready to run.
            lane.placeholder_slot = None
            modified = True
    return modified
